### PR TITLE
Code Quality: ThemedIcon, Changed IsToggle to ToggleBehavior

### PR DESCRIPTION
**Pull request by mdtauk**
**Resolved / Related Issues**

To prevent extra work, all changes to the Files codebase must link to an approved issue marked as `Ready to build`. Please insert the issue number following the hashtag with the issue number that this Pull Request resolves.
- Closes #

**Steps used to test these changes**

Stability is a top priority for Files and all changes are required to go through testing before being merged into the repo. Please include a list of steps that you used to test this PR.

1. Opened Files ...
2. ...

### CODE REVIEW

**Comment by mdtauk** [comment]
I have marked this as ready for review, I have updated the examples page now the Hierarchal paging has been merged into main.

**Comment by 0x5bfa** [comment]
- [x] Test UI is good
- [x] XAML/C# format is good
- [x] The ThemedIcon is simplified as much as possible
- [x] Refactored the infra

✅All good now, ready for review and merge.

**Comment by 0x5bfa** [comment]
@marcelwgn @yaira2

Details of changes:
- Formatted C# with tabs
- Moved Enum classes to Data folder
- Moved owner hooking up routines to another file
- Replaced IsToggled with ToggleBehavior
  - Auto: Respects both owner's toggle state and itself state
  - On: Toggle anyways
  - Off: Respect only itself's state (would be useful in Toolbar PR, and is the reason to made this PR)

**Comment by 0x5bfa** [comment]
@marcelwgn @yaira2

Details of changes:
- Formatted C# with tabs
- Moved Enum classes to Data folder
- Moved owner hooking up routines to another file
- Replaced IsToggled with ToggleBehavior
  - Auto: Respects both owner's toggle state (IsChecked) and ThemedIcon's state
  - On: Toggle anyways
  - Off: Respect only the ThemedIcon's state (e.g. explicit IsEnabled setting) (would be useful in Toolbar PR, and is the reason to made this PR)

File: src/Files.App.Controls/ThemedIcon/Data/ThemedIconToggleBehaviors.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+{
+	/// <summary>
+	/// Defines IconTypes for <see cref="ThemedIcon"/>.
+	/// </summary>
+	public enum ToggleBehaviors
+	{
+		/// <summary>
+		/// Toggle Behavior type of <see cref="ThemedIcon"/> is Auto.
+		/// </summary>
+		Auto,
```
Comment (marcelwgn) [inline src/Files.App.Controls/ThemedIcon/Data/ThemedIconToggleBehaviors.cs:14]: I'm a bit puzzled what the auto behavior is supposed to do.

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Consts.cs [hunk 2]
```diff
@@ -31,6 +31,9 @@ namespace Files.App.Controls
     [TemplateVisualState(Name = NeutralBGStateName, GroupName = IconLayerColorStateGroupName)]
     [TemplateVisualState(Name = CustomColorBGStateName, GroupName = IconLayerColorStateGroupName)]
 
+
+
+
```
Comment (marcelwgn) [inline src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Consts.cs:36]: Intentional?

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Owner.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) 2024 Files Community
```
Comment (marcelwgn) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Owner.cs:1]: This is essentially handling toggle state of the owning control, is that correct?

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs [hunk 4]
```diff
@@ -6,371 +6,224 @@
...
+	public partial class ThemedIcon : Control
+	{
+		public ThemedIcon()
+		{
+			DefaultStyleKey = typeof(ThemedIcon);
+		}
+
+		protected override void OnApplyTemplate()
+		{
+			IsEnabledChanged -= OnIsEnabledChanged;
```
Comment (marcelwgn) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:26]: Why are we unregistering before applying template?

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs [hunk 5]
```diff
@@ -6,371 +6,224 @@
...
+			OnIconTypeChanged();
+			OnIconColorTypeChanged();
+		}
+
+		// Updates paths and layers
+
+		private void OnFilledIconChanged()
+		{
+			// Updates Filled Icon from Path Data
+			if (GetTemplateChild(FilledPathIconViewBox) is not Viewbox filledViewBox)
```
Comment (marcelwgn) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:48]: I'm curious how often we are querying the Visual Tree (not only here but in general in this file) and if there might be a benefit in having a reference to these parts. That's what most of the WinUI controls do too.

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs [hunk 6]
```diff
@@ -6,371 +6,224 @@
...
+				{
+					// Toggle
+					VisualStateManager.GoToState(this, ToggleStateName, true);
+				}
+				else
+				{
+					// Use colorful ones
+					VisualStateManager.GoToState(
+						this,
+						IconColorType switch
```
Comment (marcelwgn) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:169]: Nice use of switch statement!

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Consts.cs (hunk 2 above)
Comment (marcelwgn) [inline src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Consts.cs:36]: Intentional?

File: src/Files.App.Controls/ThemedIcon/Data/ThemedIconToggleBehaviors.cs (hunk 1 above)
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/Data/ThemedIconToggleBehaviors.cs:14]: At present, when **Auto** is set, if the ThemedIcon is within a ToggleButton, that "owner" control's IsChecked state will set the Icon into the correct Toggled and Filled states.

But if we want to control it differently, we need a way to manually set this, so we check if the **ToggleBehavior** is Auto, and respond to it.  ToggleBehavior can also be On and Off, to manually set the states.

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs (hunk 4 above)
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:26]: I have seen other controls handle their event registry in this way, I assume its to have a fresh state when the Template is applied

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Owner.cs (hunk 3 above)
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.Owner.cs:1]: @0x5bfa wanted to simplify the **ThemedIcon.cs** code, and so separated this into its own partial file.  It was in that file previously

Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.Owner.cs:1]: > This is essentially handling toggle state of the owning control, is that correct?

Yes. IsToggle change for ToggleButton and AppToggleButton class and IsEnabled change for Control class.

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs (hunk 5 above)
Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:48]: Do you think we should query them one time when applying template?

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Consts.cs (hunk 2 above)
Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Consts.cs:36]: No, let me remove them.

File: src/Files.App.Controls/ThemedIcon/Data/ThemedIconToggleBehaviors.cs (hunk 1 above)
Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/Data/ThemedIconToggleBehaviors.cs:14]: It's a mixed Enum for IgnoreOwnerToggle and IsToggled.

Auto: IgnoreOwnerToggle=true, IsToggled=null
On: IsToggled=true
Off: IgnoreOwnerToggle=false

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Owner.cs (hunk 3 above)
Comment (marcelwgn) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.Owner.cs:1]: Alright, good idea.

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs (hunk 4 above)
Comment (marcelwgn) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:26]: I don't think this is actually necessary given that the template does not have to do anything with your own event.

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs (hunk 5 above)
Comment (marcelwgn) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:48]: Yes, that's what WinUI does too for performance reason.

File: src/Files.App.Controls/ThemedIcon/Data/ThemedIconToggleBehaviors.cs (hunk 1 above)
Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/Data/ThemedIconToggleBehaviors.cs:14]: Absolutely true.
Let's change comments.

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs (hunk 4 above)
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:26]: I will remove the unregistration

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs (hunk 5 above)
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:48]: I have added a method to get the template parts and set the private fields - and then we check the if those fields are null in the code.

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Consts.cs (hunk 2 above)
Comment (yaira2) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Consts.cs]: @mdtauk

Comment (yaira2) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Consts.cs]: @0x5bfa

**Comment by yaira2** [comment]
Can you resolve the merge conflicts?

**Comment by 0x5bfa** [comment]
Done. PTAL.

**Review by yaira2 — APPROVED** [review]

**Comment by yaira2** [comment]
It seems to be an accessibility issue, do the tests successfully run locally?

**Comment by 0x5bfa** [comment]
I fixed the CI but xaml formatted failing.

**Comment by yaira2** [comment]
Let's format manually so we can merge.

**Comment by 0x5bfa** [comment]
I mean, the commit right before the last commit successfully checked XAML is formatted but the last commit not while I didn't make changes on XAML.

**Pull request merged**